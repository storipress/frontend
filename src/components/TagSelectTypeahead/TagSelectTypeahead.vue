<script lang="ts">
import type { PropType } from 'vue'
import { computed, defineComponent } from 'vue'
import { SelectTypeahead } from '@storipress/core-component'
import type { TagType } from './definition'
import type { GetTagsQuery } from '~/graphql-operations'
import { GetTagsDocument } from '~/graphql-operations'
import { useQuery } from '~/lib/apollo'

export default defineComponent({
  name: 'TagSelectTypeahead',
  components: {
    SelectTypeahead,
  },
  props: {
    type: {
      type: String as PropType<'line' | 'input'>,
      default: 'line',
      validator: (value: string) => ['line', 'input'].includes(value),
    },
    label: {
      type: String,
      required: false,
    },
    placeholder: {
      type: String,
      default: '',
    },
    modelValue: {
      type: Array as PropType<TagType[]>,
      required: true,
    },
    filterOption: {
      type: Function as PropType<(option: any) => boolean>,
    },
  },
  emits: ['update:modelValue'],
  setup(props) {
    const { result } = useQuery<GetTagsQuery>(GetTagsDocument)
    const tags = computed(() => {
      const temp = result.value?.tags ?? []
      return props.filterOption ? temp?.filter(props.filterOption) : temp
    })
    return { tags }
  },
})
</script>

<template>
  <SelectTypeahead
    :label="label"
    :model-value="modelValue"
    :default-value="modelValue"
    :items="tags"
    :placeholder="placeholder"
    option-label-prop="name"
    @update:model-value="$emit('update:modelValue', $event)"
  >
    <template #default="{ item, boldMatchText }">
      <span v-html="boldMatchText(item?.name)"></span>
    </template>
  </SelectTypeahead>
</template>

<style lang="scss" scoped>
:deep.typeahead {
  & > .typeahead-wrap {
    width: 100%;
    @apply w-full;
  }
  & .simple-typeahead-list {
    @apply z-50;
  }
}
</style>
