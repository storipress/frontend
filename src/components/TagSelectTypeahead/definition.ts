import type { GetTagsQuery } from '~/graphql-operations'

export type TagType = GetTagsQuery['tags'][number]
