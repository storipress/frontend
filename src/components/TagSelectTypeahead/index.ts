export { default } from './TagSelectTypeahead.vue'
export * from './definition'
