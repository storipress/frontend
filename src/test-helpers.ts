import { NotificationsPlugin } from '@storipress/core-component'
import { createTestingPinia } from '@pinia/testing'
import { ApolloClients, provideApolloClients } from '@vue/apollo-composable'
import type { RenderOptions, RenderResult } from '@testing-library/vue'
import generatedRoutes from 'virtual:generated-pages'
import { render as vueRender } from '@testing-library/vue'
import type { RouteRecordRaw, Router } from 'vue-router'
import { createRouter, createWebHashHistory } from 'vue-router'
import { defu } from 'defu'
import { defineComponent, h } from 'vue'
import { createHead } from '@unhead/vue'
import { plugin as Slicksort } from 'vue-slicksort'
import type { TypedDocumentNode } from '@apollo/client/core'
import type { PartialDeep } from 'type-fest'
import tailwindcss from 'tailwindcss'
import postcss from 'postcss'
import type { GraphQLError } from 'graphql'
import { setActivePinia } from 'pinia'
import { server } from '../test/server'
import { KEY } from '../test/setupFiles/constant'
import type { MockHandler } from './mocks/define-graphql-mock'
import { defineGraphQLMock } from './mocks/define-graphql-mock'
import { key as confirmModelKey } from '~/components/ConfirmModalProvider/definition'
import { enableFeatures as _enableFeatures } from '~/lib/feature-flag'
import { clients } from '~/lib/apollo'

interface CustomRenderOptions extends RenderOptions {
  routes?: boolean | RouteRecordRaw[]
  stubActions?: boolean
}

interface CustomRenderResult extends RenderResult {
  router: ReturnType<typeof createRouter>
}

export const mockAxiomIngest = vi.fn()

export function render(
  // skipcq: JS-0323
  component: any,
  options: CustomRenderOptions & { css: true },
): Promise<CustomRenderResult> & CustomRenderResult
// skipcq: JS-0323
export function render(component: any, options?: CustomRenderOptions & { css?: false }): CustomRenderResult
// skipcq: JS-0323
export function render(component: any, options?: CustomRenderOptions & { css?: boolean }) {
  const routes = options?.routes
    ? options.routes === true
      ? generatedRoutes
      : options.routes
    : [
        { path: '/', component: defineComponent({ setup: () => () => h('div', 'Home') }) },
        { path: '/**', component: defineComponent({ setup: () => () => h('div', 'CatchAll') }) },
      ]
  const router = createRouter({
    history: createWebHashHistory(),
    routes,
  })

  const renderOptions: RenderOptions = {
    ...options,
    global: {
      ...options?.global,
      plugins: [
        ...(options?.global?.plugins ?? []),
        NotificationsPlugin,
        Slicksort,
        setupTestPinia(options?.stubActions),
        createHead(),
        router,
        (app) => {
          app.provide(KEY, 'secret-key')
          app.provide(ApolloClients, clients)
          app.provide(confirmModelKey, { confirm: vi.fn(), modalList: [] })
        },
      ],
    },
  }
  const res = vueRender(component, renderOptions)

  // when css is true, create a new instance of tailwindcss and postcss
  if (options?.css) {
    const lazyResult = postcss([
      tailwindcss({
        content: [{ raw: res.container.innerHTML }],
      }),
    ]).process('@tailwind utilities;', { from: undefined })
    const promise: Promise<RenderResult> & RenderResult = lazyResult.then((result) => {
      const style = document.createElement('style')
      style.innerHTML = result.css
      res.container.append(style)
      return {
        ...res,
        router,
      }
    }) as Promise<CustomRenderResult> & CustomRenderResult

    return new Proxy(promise, {
      get(target, key) {
        if (key === 'then') {
          const then = target.then.bind(target)
          return (...args: Parameters<typeof then>) => {
            then(...args)
          }
        }
        throw new Error('Please await the returned promise if you are using `css: true`')
      },
    })
  }

  return {
    ...res,
    router,
  }
}

interface Composable<Params extends unknown[], Return> {
  (...params: Params): Return
}

export function runComposable<Params extends unknown[], Return>(
  composable: Composable<Params, Return>,
  ...params: Params
): { result: Return; router: Router } {
  const router = createRouter({
    history: createWebHashHistory(),
    routes: [{ path: '/', component: defineComponent({ setup: () => () => h('div', 'Home') }) }],
  })

  const app = createApp({
    setup: () => () => h('div'),
  })

  app.use(NotificationsPlugin)
  app.use(Slicksort)
  app.use(setupTestPinia())
  app.use(createHead())
  app.use(router)
  app.provide(KEY, 'secret-key')
  app.provide(ApolloClients, clients)
  app.provide(confirmModelKey, { confirm: vi.fn(), modalList: [] })

  return {
    result: app.runWithContext(() => composable(...params)),
    router,
  }
}

export function setupApolloClient() {
  provideApolloClients(
    new Proxy(
      {},
      {
        get: () => {
          return clients.default
        },
      },
    ),
  )
  resetApolloCache()
}

export function resetApolloCache() {
  clients.default.cache.reset({})
}

export function setupTestPinia(stubActions?: boolean) {
  return createTestingPinia({
    stubActions,
    initialState: {
      client: {
        clientID: 'D6RX98VXN',
        defaultClientID: 'default',
        token: 'cAA6LoWllmqdXaYYid1AxpEDy5liOBGp5ppKPNKM',
        isResetPassword: false,
      },
      'search-condition': {
        text: undefined,
        people: undefined,
        tags: undefined,
        range: undefined,
      },
      site: {
        site: {
          name: "Joe's Blog",
          timezone: 'Asia/Taipei',
          email: 'test@storipress.com',
          socials: '{"Facebook":"www.facebook.com\\/Storipress"}',
        },
        siteTutorials: {
          setSocialConnect: true,
          setPublicationDetail: true,
          setCreateDesks: true,
          setDomain: true,
        },
      },
      me: {
        me: {
          id: 151,
          role: 'owner',
        },
      },
      integration: {
        integrations: [
          { activated_at: null, data: '{"shortname":null}', key: 'disqus', order: 6 },
          {
            activated_at: null,
            data: '[]',
            key: 'webflow',
            order: 7,
          },
        ],
      },
      knownErrors: {
        errors: [],
      },
    },
  })
}

export function setupTestEnv() {
  setActivePinia(setupTestPinia())
  setupApolloClient()
}

export function useTestEnv() {
  beforeEach(setupTestEnv)
}

export function useFakeTimer(now?: number | Date) {
  beforeEach(() => {
    vi.useFakeTimers({ now })
  })

  afterEach(() => {
    vi.useRealTimers()
  })
}

export function mockResponse<QueryOutput extends Record<string, any>>(
  doc: TypedDocumentNode<QueryOutput, any>,
  data: QueryOutput,
) {
  const handler = defineGraphQLMock(doc, data, (output, res, ctx) => res.once(ctx.data(output)))

  server.use(handler)
}

export function mockResponseOnce<QueryOutput extends Record<string, any>, Variables extends Record<string, any>>(
  originalHandler: MockHandler<QueryOutput, Variables>,
  patchedQuery: PartialDeep<QueryOutput> = {} as PartialDeep<QueryOutput>,
) {
  const handler = defineGraphQLMock<QueryOutput, Variables>(
    originalHandler.doc,
    (req) => {
      return defu(patchedQuery, originalHandler.resolveOutput(req)) as QueryOutput
    },
    (output, res, ctx) => res.once(ctx.data(output)),
  )

  server.use(handler)
}

export function mockResponseError<QueryOutput extends Record<string, any>, Variables extends Record<string, any>>(
  originalHandler: MockHandler<QueryOutput, Variables>,
  errors: Partial<GraphQLError>[],
) {
  const handler = defineGraphQLMock<QueryOutput, Variables>(
    originalHandler.doc,
    () => {
      return {} as QueryOutput
    },
    (_output, res, ctx) => res.once(ctx.errors(errors)),
  )

  server.use(handler)
}

export function enableFeatures(features: string[]) {
  _enableFeatures(features)
}
