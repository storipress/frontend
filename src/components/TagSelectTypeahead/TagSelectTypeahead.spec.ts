/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import { fireEvent, waitFor, within } from '@testing-library/vue'
import { expect, it } from 'vitest'
import { html } from 'proper-tags'
import TagSelectTypeahead from './TagSelectTypeahead.vue'
import { render } from '~/test-helpers'

it('options list can trigger', async () => {
  const { container, getByLabelText, getByText, queryByText } = render(TagSelectTypeahead, {
    props: {
      modelValue: [],
      placeholder: 'Tags',
    },
  })
  const input = getByLabelText(/Tags/)

  await input.focus()
  await waitFor(() => {
    expect(container.querySelector('.simple-typeahead-list')).toBeVisible()
  })

  const listItems = Array.from(container.querySelectorAll('.simple-typeahead-list-item'))
  expect(getByText('World')).toBeVisible()
  expect(listItems.map((item) => item.textContent)).toMatchSnapshot()

  await input.blur()
  expect(queryByText('World')).not.toBeInTheDocument()
})

/*
 * In core-component/SelectTypeahead components, `selected` is in the same `<ul />` as the `option list`.
 * So when the option list is enabled, at least one set of options will be rendered.
 */
it('item selected', async () => {
  const { container, getByLabelText, getByText, queryAllByText, getByRole, queryByText } = render({
    components: { TagSelectTypeahead },
    setup: () => ({ tagsList: ref([]) }),
    template: html`<tag-select-typeahead class="grow" placeholder="Tags" v-model="tagsList" />`,
  })

  const input = getByLabelText(/Tags/)
  await input.focus()
  await waitFor(() => {
    expect(container.querySelector('.simple-typeahead-list')).toBeVisible()
  })

  expect(getByText('World')).toBeVisible()

  const optionsContainer: HTMLElement =
    container.querySelector('.simple-typeahead-list') || document.createElement('div')

  const option = within(optionsContainer).getByText(/World/i)
  await fireEvent.click(option)
  await input.blur()

  expect(queryAllByText('World')).toHaveLength(1)

  const crossButton = getByRole('button', { name: 'remove' })
  await fireEvent.click(crossButton)

  expect(queryByText('World')).not.toBeInTheDocument()
})

it('data filter', async () => {
  const { container, getByLabelText, queryByText } = render(TagSelectTypeahead, {
    props: {
      modelValue: [],
      placeholder: 'Tags',
      filterOption: ({ name }: { name: string }) => {
        if (/World/i.test(name)) return false
        return true
      },
    },
  })
  const input = getByLabelText(/Tags/)

  await input.focus()
  await waitFor(() => {
    expect(container.querySelector('.simple-typeahead-list')).toBeVisible()
  })

  expect(queryByText('World')).not.toBeInTheDocument()
})
