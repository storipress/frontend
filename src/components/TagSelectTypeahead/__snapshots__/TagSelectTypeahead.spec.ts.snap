// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options list can trigger 1`] = `
[
  "World",
  "Business",
  "Opinion",
  "Tech",
  "Science",
  "Health",
  "Sports",
  "Arts",
  "Books",
  "Style",
  "Food",
  "Video",
]
`;
